File: cPlusPlusCode/findContours.cpp
#include<opencv2/imgproc.hpp>
#include<opencv2/highgui.hpp>
#include<iostream>

using namespace cv;
using namespace std;

int main(){

    //Read input image in gray scale
    Mat image = imread("../assets/anish.jpg", IMREAD_GRAYSCALE);

    //Check if image exits
    if(image.empty()){
        cout<<"can not find image"<<endl;
        return 0;
    }

    Mat thresholdedImage;
    //Apply canny to the input image
    Canny(image, thresholdedImage, 50, 150, 3);

    //To store contours
    vector<vector<Point>> contours;
    
    //To store hierarchy(nestedness)
    vector<Vec4i> hierarchy;

    //Find contours
    findContours(thresholdedImage, contours, hierarchy, RETR_TREE, CHAIN_APPROX_SIMPLE);

    //Output image to draw contours on
    Mat output;
    output = Mat::zeros( thresholdedImage.size(), CV_8UC3 );

    RNG rng(12345);
    
    // Draw contours.
    for( size_t i = 0; i< contours.size(); i++ ){
        Scalar color = Scalar( rng.uniform(0, 255), rng.uniform(0,255), rng.uniform(0,255) );
        drawContours(output, contours, i, color, 2);
    }

    //create windows to display images
    namedWindow("image", WINDOW_AUTOSIZE);
    namedWindow("canny", WINDOW_AUTOSIZE);
    namedWindow("contours", WINDOW_AUTOSIZE);
    
    /// display images
    imshow("image", image);
    imshow("canny", thresholdedImage);
    imshow("contours", output );

    //Press esc to exit the program
    waitKey(0);

    //close all the opened windows
    destroyAllWindows();

    return 0;
}

File: pythonCode/findContours.py
import cv2
import sys
import numpy as np
import random

#read input image
image = cv2.imread("../assets/anish.jpg", cv2.IMREAD_GRAYSCALE)

#check if image exists
if image is None:
    print("can not find image")
    sys.exit()

#apply canny to the input image
canny = cv2.Canny(image, 50, 150, apertureSize=3)

#find contours
temp, contours, hierarchy = cv2.findContours(canny, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)

#output image to draw contours on
output = np.zeros((image.shape[0], image.shape[1], 3), dtype=np.uint8)

#draw contours
for i in range(0, len(contours)):
    cv2.drawContours(output, contours, i, (random.randint(0, 255), random.randint(0, 255), random.randint(0, 255)), 2)


#create windows to display images
cv2.namedWindow("image", cv2.WINDOW_AUTOSIZE)
cv2.namedWindow("canny", cv2.WINDOW_AUTOSIZE)
cv2.namedWindow("contours", cv2.WINDOW_AUTOSIZE)

#display images
cv2.imshow("image", image)
cv2.imshow("canny", canny)
cv2.imshow("contours", output)

#press esc to exit the program
cv2.waitKey(0)

#close all the opened windows
cv2.destroyAllWindows()